NMS V4 to ONNX NMS optimizaion.

ayermolo opened the pull request:
Summary:
Optimization for some flavors of SSD converted from TF. There is a family of SSD networks that use different core networks, but have the same post processing.

Network use NMS V4 which can only handle 1 class. Which results in 80 NMS nodes, with all the corresponding sub graphs to split the data then filter results and concat everything together. It also uses TopK to pack all results together.
Optimization Isolates NMS V4 subgraph and replaces with ONNX NMS and appropriate subgraph. Due to being able to handle multiple classes in ONNX NMS the first TopK also becomes redundant.

Optimization finds NMS nodes and isolates the divergence and convergence points and replaces it with a new much smaller sub graph that uses ONNX NMS.

Using utility functions in Utils.h created by @ksaurabh-cadence 
Documentation:
Before optimization:
![SSD_TF_graph](https://user-images.githubusercontent.com/43973793/81863589-5376ad00-9520-11ea-9dfb-b0599d885a79.jpg)

After optimization:
![SSD_TF_graph_to_onnx_nms](https://user-images.githubusercontent.com/43973793/81863593-54a7da00-9520-11ea-9230-948008643f77.jpg)
[Optional Fixes #issue]

Test Plan:
Unit Tests, TF SSD networks.
Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

ayermolo commented:
@mciprian13 @jfix71 Any interest in this? Otherwise I'll just close it.

ayermolo commented on include/glow/Graph/Utils.h:
```diff
@@ -132,6 +132,67 @@ class GraphPreOrderVisitor : public PreOrderVisitor {
   }
 };
 
+using Edge = std::pair<Node *, NodeValue>;
```
I think original intent was to capture the upward relationship in case of both up and down traversal, and make it easy to access the "destination" node itself by code that uses traverseUp/traverseDown. Also just having input number would complicate down traversal code a bit as we have to figure out from user node which of it's input corresponds to def node.
Vs now we can just do this:

```
Node *childNode = user.getUser();
    if (matchCondition(childNode)) {
      edgeList.push_back(Edge(childNode, currNodeVal));
    }
```

@ksaurabh-cadence  thoughts?

ksaurabh-cadence replied:
IMHO NodeValue is still light weight. As @ayermolo explained this interpretation was simpler to visualize/write traversal in either direction and provided more direct access to the connecting node.

ayermolo commented:
@jfix71 Thanks. Rebased on latest, and addressed the comments.
Seems like failure is unrelated: ThreadPoolExecutorTest.ConcurrentMultiNode.
Couldn't reproduce it locally, or how my optimization would affect it.

ayermolo commented:
@ksaurabh-cadence @aturetsk Since I am no longer on a project can you take over, or close?

Thanks

ksaurabh-cadence commented:
@ayermolo sure!

@mortzur Can you please take a look?
